### ISSUE TITLE
[Question] Precision of bfloat16 computations

### ISSUE THREAD

**Issue by lukasHoel** [pytorch-labs/attention-gym#131]
Hi,

I notice that in `bfloat16` mode, the results of sdpa vs flex attention vary quite a bit.
Is this expected behavior due to the low precision (only 7 bit for mantissa) or should the results actually be much closer?
Below I have chosen q,k,v tensors such that the dimension and value ranges match that of a real network I currently use.
I am worried that mixed-precision training with flex attention might be worse because of less accurate computations?

```
bs = 1
n_heads = 16
seq_len = 240
head_dim = 128
flex_attn_fn = torch.compile(flex_attention)
torch.manual_seed(0)
scale = 10
q = torch.randn(bs, n_heads, seq_len, head_dim).cuda() * scale
k = torch.randn(bs, n_heads, seq_len, head_dim).cuda() * scale
v = torch.randn(bs, n_heads, seq_len, head_dim).cuda() * scale
for dtype in [torch.float32, torch.bfloat16, torch.float16]:
    with torch.autocast(device_type="cuda", dtype=dtype):
        sdpa_out = torch.nn.functional.scaled_dot_product_attention(q, k, v)
        flex_attn_out = flex_attn_fn(q.to(dtype), k.to(dtype), v.to(dtype))
        print(dtype, torch.allclose(sdpa_out, flex_attn_out, atol=1e-3), (sdpa_out - flex_attn_out).abs().max().item())
```

Output :
```
torch.float32 False 0.001984238624572754
torch.bfloat16 False 0.125
torch.float16 False 0.03125
```

Computed results with `torch 2.6.0+cu118` and the latest nightly release `2.8.0.dev20250404+cu118`.

**Comment by drisspg** [comment]
```Python

import torch
import torch.nn.functional as F
from torch.nn.attention.flex_attention import flex_attention

flex_compiled = torch.compile(flex_attention)


def eager_attention(q, k, v):
    scores = q @ k.transpose(-2, -1) / q.shape[-1] ** 0.5
    return scores.softmax(dim=-1) @ v


torch.manual_seed(0)
bs, n_heads, seq_len, head_dim, scale = 1, 16, 240, 128, 10
q, k, v = (torch.randn(bs, n_heads, seq_len, head_dim, device="cuda") * scale for _ in range(3))

print(f"{'dtype':<16} {'flex_vs_sdpa':>12} {'flex_err':>12} {'sdpa_err':>12} {'eager_err':>12}")
for dtype in [torch.float32, torch.bfloat16, torch.float16]:
    qd, kd, vd = q.to(dtype), k.to(dtype), v.to(dtype)
    ref = eager_attention(qd.double(), kd.double(), vd.double())
    outs = {
        "flex": flex_compiled(qd, kd, vd),
        "sdpa": F.scaled_dot_product_attention(qd, kd, vd),
        "eager": eager_attention(qd, kd, vd),
    }
    errs = {name: (out.double() - ref).abs().max().item() for name, out in outs.items()}
    flex_vs_sdpa = (outs["flex"] - outs["sdpa"]).abs().max().item()
    print(
        f"{str(dtype):<16} {flex_vs_sdpa:>12.2e}"
        f" {errs['flex']:>12.2e} {errs['sdpa']:>12.2e} {errs['eager']:>12.2e}"
    )

```

I think your example just shows the limitations of FP precision ^^ produces:

```
dtype            flex_vs_sdpa     flex_err     sdpa_err    eager_err
torch.float32        1.90e-03     1.81e-03     8.47e-04     1.87e-03
torch.bfloat16       1.25e-01     1.21e-01     1.33e-01     1.91e+01
torch.float16        1.56e-02     1.54e-02     1.85e-02     2.76e+00
```

eager err is worse since it materializes softmax scores in lower precison
